pepr/privacy/art_inference_wrapper.py

```diff
@@ -0,0 +1,974 @@
+        * attack_model: (optional) The attack model to train, optional. If none is
+          provided, a default model will be created.
+
+    data : numpy.ndarray
+        Dataset with all input images used to attack the target models.
+    labels : numpy.ndarray
+        Array of all labels used to attack the target models.
+    data_conf : dict
+        Dictionary describing for every target model which record-indices should be used
+        for the attack.
+
+        * train_record_indices (np.ndarray): Input to training process. Includes all
+          features used to train the original model.
+        * test_record_indices (np.ndarray): Test records that are not used in the
+          training of the target model.
+        * attack_record_indices (np.ndarray): Input to attack. Includes all features
+          except the attacked feature.
+        * attack_membership_status (np.ndarray): True membership status of the
+          attack_records 1 indicates a member and 0 indicates non-member. This is used
+          to compare the attacks results with the true membership status.
+
+    target_models : iterable
+        List of target models which should be tested.
+    """
+
+    def __init__(
+        self, attack_alias, attack_pars, data, labels, data_conf, target_models
+    ):
+        pars_descriptors = {
+            "attack_model_type": "Attack model type",
+            "input_type": "Input type",
+            "attack_model": "Attack model",
+        }
+
+        # Handle specific attack class parameters
+        params = {}
+        for k in pars_descriptors:
+            if k in attack_pars:
+                params[k] = attack_pars[k]
+
+        inference_attacks = []
+        for target_model in target_models:
+            target_classifier = KerasClassifier(target_model, clip_values=(0, 1))
+            inference_attacks.append(
+                membership_inference.MembershipInferenceBlackBox(
+                    classifier=target_classifier, **params
+                )
+            )
+
+        super().__init__(
+            attack_alias,
+            data,
+            labels,
+            data_conf,
+            target_models,
+            inference_attacks,
+            pars_descriptors,
+        )
+
+        self.report_section = report.ReportSection(
+            "Membership Inference Black-Box",
+            self.attack_alias,
+            "ART_MembershipInferenceBlackBox",
+        )
+
+    def __getstate__(self):
+        if self.inference_attacks[0].default_model:
+            for inference_attack in self.inference_attacks:
+                # Delete attack model, because some are not pickable
+                del inference_attack.__dict__["attack_model"]
+        return super().__getstate__()
+
+    def __setstate__(self, state):
+        super().__setstate__(state)
+        # TODO: Restore attack models (not required for just generating report)
```

Comment:
What's about this Todo? 👀

Reply:
This todo is for whenever we want to be able to resume attacks saved on the disk. 

The `attack_model` attribute of the object must be stored and restored separately due to the incompatibility to Pythons's pickle. This can be done with `model.save()` (TensorFlow) or `torch.save()` (PyTorch). But I think this is currently not required.

Reply:
Can you add the second paragraph from you answer as an explanation to the Todo?

Reply:
Added in 2cd28cb